gcloud-java-compute/src/test/java/com/google/gcloud/compute/OperationIdTest.java

```diff
@@ -0,0 +1,143 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gcloud.compute;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertTrue;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class OperationIdTest {
+
+  private static final String PROJECT = "project";
+  private static final String ZONE = "zone";
+  private static final String REGION = "region";
+  private static final String NAME = "op";
+  private static final String GLOBAL_URL =
+      "https://www.googleapis.com/compute/v1/projects/project/global/operations/op";
+  private static final String ZONE_URL =
+      "https://www.googleapis.com/compute/v1/projects/project/zones/zone/operations/op";
+  private static final String REGION_URL =
+      "https://www.googleapis.com/compute/v1/projects/project/regions/region/operations/op";
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  @Test
+  public void testOf() {
+    GlobalOperationId operationId = GlobalOperationId.of(PROJECT, NAME);
+    assertEquals(PROJECT, operationId.project());
+    assertEquals(NAME, operationId.operation());
+    assertEquals(GLOBAL_URL, operationId.selfLink());
+    operationId = GlobalOperationId.of(NAME);
+    assertNull(operationId.project());
+    assertEquals(NAME, operationId.operation());
+    ZoneOperationId zoneOperationId = ZoneOperationId.of(PROJECT, ZONE, NAME);
+    assertEquals(PROJECT, zoneOperationId.project());
+    assertEquals(ZONE, zoneOperationId.zone());
+    assertEquals(NAME, zoneOperationId.operation());
+    assertEquals(ZONE_URL, zoneOperationId.selfLink());
+    zoneOperationId = ZoneOperationId.of(ZONE, NAME);
+    assertNull(zoneOperationId.project());
+    assertEquals(ZONE, zoneOperationId.zone());
+    assertEquals(NAME, zoneOperationId.operation());
+    RegionOperationId regionOperationId = RegionOperationId.of(PROJECT, REGION, NAME);
+    assertEquals(PROJECT, regionOperationId.project());
+    assertEquals(REGION, regionOperationId.region());
+    assertEquals(NAME, regionOperationId.operation());
+    assertEquals(REGION_URL, regionOperationId.selfLink());
+    regionOperationId = RegionOperationId.of(REGION, NAME);
+    assertNull(regionOperationId.project());
+    assertEquals(REGION, regionOperationId.region());
+    assertEquals(NAME, regionOperationId.operation());
+  }
+
+  @Test
+  public void testToAndFromUrl() {
+    GlobalOperationId operationId = GlobalOperationId.of(PROJECT, NAME);
+    compareOperationId(operationId, GlobalOperationId.fromUrl(operationId.selfLink()));
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("notMatchingUrl is not a valid global operation URL");
```

Comment:
I'm fairly sure that once an exception is hit, this method doesn't execute any other code below.  Perhaps we should split this method up into three, one per type of identity?